src/prefect/server/database/configurations.py

```diff
@@ -246,32 +268,78 @@ async def dispose_engine(cache_key):
+        # even with the WAL we can get busy errors if we have transactions that:
+        #  - t1 reads from a database
+        #  - t2 inserts to the database
+        #  - t1 tries to insert to the database
+        # this can be resolved by using the IMMEDIATE transaction mode in t1
+        cursor.execute("PRAGMA journal_mode = WAL;")
+
         # enable foreign keys
-        conn.execute(sa.text("PRAGMA foreign_keys = ON;"))
+        cursor.execute("PRAGMA foreign_keys = ON;")
 
-        # write to a write-ahead-log instead and regularly
-        # commit the changes
-        # this allows multiple concurrent readers even during
-        # a write transaction
-        conn.execute(sa.text("PRAGMA journal_mode = WAL;"))
+        # disable legacy alter table behavior as it will cause problems during
+        # migrations when tables are renamed as references would otherwise be retained
+        # in some locations
+        # https://www.sqlite.org/pragma.html#pragma_legacy_alter_table
+        cursor.execute("PRAGMA legacy_alter_table=OFF")
 
         # when using the WAL, we do need to sync changes on every write. sqlite
         # recommends using 'normal' mode which is much faster
-        conn.execute(sa.text("PRAGMA synchronous = NORMAL;"))
+        cursor.execute("PRAGMA synchronous = NORMAL;")
 
         # a higher cache size (default of 2000) for more aggressive performance
-        conn.execute(sa.text("PRAGMA cache_size = 20000;"))
+        cursor.execute("PRAGMA cache_size = 20000;")
 
         # wait for this amount of time while a table is locked
         # before returning and rasing an error
         # setting the value very high allows for more 'concurrency'
         # without running into errors, but may result in slow api calls
-        conn.execute(sa.text("PRAGMA busy_timeout = 60000;"))  # 60s
-        conn.commit()
+        cursor.execute("PRAGMA busy_timeout = 60000;")  # 60s
+
+        cursor.close()
+
+    def begin_sqlite_conn(self, conn, record):
+        # disable pysqlite's emitting of the BEGIN statement entirely.
+        # also stops it from emitting COMMIT before any DDL.
+        # requires `begin_sqlite_stmt`
+        # see https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#serializable-isolation-savepoints-transactional-ddl
+        conn.isolation_level = None
+
+    def begin_sqlite_stmt(self, conn):
+        # emit our own BEGIN
+        # requires `begin_sqlite_conn`
+        # see https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#serializable-isolation-savepoints-transactional-ddl
+        mode = SQLITE_BEGIN_MODE.get()
+        if mode is not None:
+            conn.exec_driver_sql(f"BEGIN {mode}")
+
+        # Note this is intentionally a no-op if there is no BEGIN MODE set
+        # This allows us to use SQLite's default behavior for reads which do not need
+        # to be wrapped in a long-running transaction
```

Comment:
This is a little dangerous because if someone begins a transaction on a session manually we will now ignore it. This matches the previous behavior of pysqlite so it shouldn't necessarily be a regression but I'm worried about the possibility of future footguns.